index.js

```diff
@@ -199,10 +203,63 @@ function session(options){
 
       setcookie(res, name, req.sessionID, secrets[0], cookie.data);
     });
+    
+    var _write = res.write;
+
+    if (saveBeforeLocationHeader) {
+      var _writeHead = res.writeHead;
+      var headWritten = false;
+      res.writeHead = function writeHead(statusCode, reason, obj) {
+        if (headWritten) return;
+        headWritten = true;
 
+        // search for location header only for 3xx status codes
+        if (statusCode < 300 || statusCode >= 400) {
+          return _writeHead.call(res, statusCode, reason, obj);
+        }
+        
+        this.statusCode = statusCode;
+        
+        if ('string' != typeof(reason)) {
+          obj = reason;
+          reason = undefined;
+        }
+        
+        if (obj) {
+          var keys = Object.keys(obj);
+          for (var i = 0, l = keys.length; i < l; i++) {
+            var k = keys[i];
+            if (k) res.setHeader(k, obj[k]);
+          }
+        }
+        
+        // we have a `location` header so we must buffer all writes
+        if (res.getHeader('location') != null) {
+          debug('redirect found, buffering response')
+          buffered = true;
+          bufferedChunks = new Buffer(0);
+          writeHeadLater = function() { _writeHead.call(res, statusCode, reason); };
+        } else {
+          _writeHead.call(res, statusCode, reason);
+        }
+      };
+      
+      var __write = _write;
+      res.write = _write = function write(chunk, encoding, callback) {
+        if (!headWritten) res.writeHead(this.statusCode);
+        
+        if (buffered) {
+          bufferedChunks = Buffer.concat([bufferedChunks, !Buffer.isBuffer(chunk) ? new Buffer(chunk, encoding) : chunk]);
+          if ('function' === typeof(callback)) setImmediate(callback);
```

Comment:
The `callback` should only be called once we actually performed the write, so these need to be buffered up and called once we write out our buffer.